dvc/repo/data.py

```diff
@@ -127,17 +111,16 @@ def _diff(
     new_oid: Optional["HashInfo"],
     new_obj: Optional["HashFile"],
     odb: "HashFileDB",
-    with_dirs: bool = False,
     granular: bool = False,
 ) -> Dict[str, List[str]]:
     if not granular:
         return _shallow_diff(root, old_oid, new_oid, odb)
     if (old_oid and not old_obj) or (new_oid and not new_obj):
         # we don't have enough information to give full details
-        unknown = _get_obj_items(root, new_obj or old_obj)
+        unknown = _get_obj_items(root, new_obj)
```

Comment:
This change also makes `unknown` to return list of files from newer side only.
i.e. `diff(HEAD, index)`  will only list unknown from `index`, and only if cache on index side exists. If there is no cache in HEAD side, it's not reported.

Similarly, for `diff(index, workspace)`, this will only return unknowns from `workspace` side. But this was always the case.